worlds/stardew_valley/logic.py

```diff
@@ -1387,10 +1389,93 @@ def can_reach_woods_depth(self, depth: int) -> StardewRule:
+        return self.has_relationship("Wizard", 3) & self.can_reach_region(SVRegion.pierre_store) & \
+            self.can_reach_region(SVRegion.wizard_tower)
+
+    def has_any_spell(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        return self.can_earn_spells()
+
+    def has_decent_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 3)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule | magic_attack_options_rule | magic_support_options_rule
+
+    def has_good_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 3)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule | magic_attack_options_rule | magic_support_options_rule
+
+    def has_great_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 6)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule & magic_attack_options_rule & magic_support_options_rule
+
+    def has_amazing_spells(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        magic_resource_rule = self.can_earn_spells() & self.received("Magic Level", 9)
+        magic_attack_options_rule = self.received("Spell: Fireball") | self.received(
+            "Spell: Frostbite") | self.received("Spell: Shockwave") | self.received("Spell: Spirit")
+        magic_support_options_rule = self.received("Spell: Heal") | self.received("Spell: Tendrils")
+        return magic_resource_rule & magic_attack_options_rule & magic_support_options_rule
+
+    def can_blink(self) -> StardewRule:
+        if ModNames.magic not in self.options[options.Mods]:
+            return False_()
+        return self.received("Spell: Blink") & self.can_earn_spells()
```

Comment:
Can blink includes can_earn_spells(), but in many cases, you manually added it as well next to the usage of can_blink. To be double checked